### PR TITLE
Cache values in UnderscoreToCamelCaseFilter and CamelCaseToUnderscoreFilter

### LINKED ISSUE #68: Cache values in UnderscoreToCamelCaseFilter and CamelCaseToUnderscoreFilter
### Feature Request
|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes/no
| BC Break    | yes/no

#### Summary
Currently the `CamelCaseToUnderscoreFilter` and the `UnderscoreToCamelCaseFilter` execute the regex on every `filter` call. I think it could improve the performance if the code gets adjusted like the following:

```php
static array $transformedFilters = [];

public function filter(string $value): string
{
    if (array_key_exists($value, static::$transformedFilters)) {
        return static::$transformedFilters[$value];
    }

    $pcreInfo = $this->getPatternAndReplacement(
    // a unicode safe way of converting characters to \x00\x00 notation
        preg_quote('_', '#')
    );

    $filtered = preg_replace_callback(
        $pcreInfo->pattern,
        $pcreInfo->replacement,
        $value
    );

    $lcFirstFunction = $this->getLcFirstFunction();
    $filteredValue = $lcFirstFunction($filtered);

    static::$transformedFilters[$value] = $filteredValue;

    return $filteredValue;
}
```

Should I create a pull request for this change?

### PR DESCRIPTION
**Pull request by DerKnerd**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
I added a field name cache to the CamelCaseToUnderscoreFilter and the UnderscoreToCamelCaseFilter. By that the performance could be improved by the factor 4.

### CODE REVIEW

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php [hunk 1]
```diff
@@ -18,10 +19,17 @@ final class UnderscoreToCamelCaseFilter
...
+
     public function filter(string $value): string
     {
+        if (array_key_exists($value, $this->transformedFilters)) {
+            return $this->transformedFilters[$value];
+        }
+
         $pcreInfo = $this->getPatternAndReplacement(
-            // a unicode safe way of converting characters to \x00\x00 notation
+        // a unicode safe way of converting characters to \x00\x00 notation
```
Comment (weierophinney) [inline src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: Previous indentation was correct.

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php [hunk 2]
```diff
@@ -58,7 +71,7 @@ function ($matches) {
             )
             : new PcreReplacement(
                 '#(' . $pregQuotedSeparator . ')'
-                    . '([^\p{Z}\p{Ll}]{1}|[a-zA-Z]{1})#u',
+                . '([^\p{Z}\p{Ll}]{1}|[a-zA-Z]{1})#u',
```
Comment (weierophinney) [inline src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: Previous indentation made it more clear that the second line was a continuation of the first argument; I'd argue we should keep it as-is.

File: composer.json [hunk 3]
```diff
@@ -26,16 +26,18 @@
     },
     "require": {
         "php": "^7.3 || ~8.0.0",
-        "webmozart/assert": "^1.10",
         "laminas/laminas-stdlib": "^3.3",
-        "laminas/laminas-zendframework-bridge": "^1.0"
+        "laminas/laminas-zendframework-bridge": "^1.0",
+        "psr/cache": "1.0.1",
+        "webmozart/assert": "^1.10"
```
Comment (weierophinney) [inline composer.json:32]: Why were these added? I don't see usage of them in the diff provided...

File: .gitignore [hunk 4]
```diff
@@ -7,3 +7,4 @@
 /laminas-mkdoc-theme/
 /phpunit.xml
 /vendor/
+.idea
```
Comment (weierophinney) [inline .gitignore]: Please revert; this is something to add to your user global git ignore rules.

**Review by weierophinney — CHANGES_REQUESTED** [review]
A few nitpicks, but one bigger question about the addition of dependencies to the package.

**Comment by weierophinney** [comment]
Also, can you provide the benchmark results both before and after applying changes? Thanks!

File: src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php [hunk 5]
```diff
@@ -16,13 +17,26 @@ final class CamelCaseToUnderscoreFilter
 {
     use StringSupportTrait;
 
+    /** @var string[] $transformedFilters */
+    private $transformedFilters = [];
```
Comment (Ocramius) [inline src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php:21]: I know you act in good faith here, but a long running process is just going to die, when/if using this component this way 😟

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: composer.json (hunk 3 above)
Comment (DerKnerd) [reply composer.json:32]: In the github action composer was not able to get an installable set of packages when running with PHP 7.4. After pinning the version of psr/cache it worked. The webmozart/assert might be not needed. Check this Github action run: https://github.com/laminas/laminas-hydrator/runs/3137326621

**Comment by DerKnerd** [comment]
I made the benchmark with the default behavior. And this is the result:

| benchmark                        | subject     | set | revs | its | mem_peak | mode        | rstdev |
----------------------------------|-------------|-----|------|-----|----------|-------------|--------|
| UnderscoreToCamelCaseFilterBench | benchFilter | 0   | 2    | 2   | 1.545mb  | 1,039.056μs | ±2.74% |

And adjusting the method this is the result:

| benchmark                        | subject     | set | revs | its | mem_peak | mode      | rstdev |
|----------------------------------|-------------|-----|------|-----|----------|-----------|--------|
| UnderscoreToCamelCaseFilterBench | benchFilter | 0   | 2    | 2   | 1.545mb  | 250.234μs | ±3.30% |

File: src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php (hunk 5 above)
Comment (DerKnerd) [reply src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php:21]: This was the whole point of the PR, to cache the filters, like it is already done with the ReflectionProperties

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php (hunk 2 above)
Comment (DerKnerd) [reply src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: I updated the code.

File: .gitignore (hunk 4 above)
Comment (DerKnerd) [reply .gitignore]: I updated the code.

File: src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php (hunk 1 above)
Comment (DerKnerd) [reply src/NamingStrategy/UnderscoreNamingStrategy/UnderscoreToCamelCaseFilter.php]: I updated the code.

File: src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php (hunk 5 above)
Comment (Ocramius) [reply src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php:21]: `ReflectionProperty[]` are a finite amount of values, because you can only have that many in memory.

the transformed strings are potentially billions/more.

Comment (DerKnerd) [reply src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php:21]: Ah I see, then you guys need to discuss the pr. It is deliberately as non static property. I thought only a static property would cause these issues.

Comment (DerKnerd) [reply src/NamingStrategy/UnderscoreNamingStrategy/CamelCaseToUnderscoreFilter.php:21]: One thing I just noticed, the amount of ReflectionProperties and transformed strings will always be the same cause the filter gets called based on the ReflectionProperty.
